lib/pos_handshake_controller.go

```diff
@@ -0,0 +1,160 @@
+package lib
+
+import (
+	"fmt"
+	"github.com/decred/dcrd/lru"
+	"github.com/golang/glog"
+	"math"
+)
+
+// HandshakeController is a structure that handles the handshake process with remote nodes. It is the entry point for
+// initiating a handshake with a remote node. It is also responsible for handling version/verack messages from remote
+// nodes. And for handling the handshake complete control message.
+type HandshakeController struct {
+	rnManager  *RemoteNodeManager
+	usedNonces lru.Cache
+}
+
+func NewHandshakeController(rnManager *RemoteNodeManager) *HandshakeController {
+
+	vm := &HandshakeController{
+		rnManager:  rnManager,
+		usedNonces: lru.NewCache(1000),
+	}
+
+	return vm
+}
+
+// InitiateHandshake kicks off handshake with a remote node.
+func (hc *HandshakeController) InitiateHandshake(rn *RemoteNode) {
+	nonce := uint64(RandInt64(math.MaxInt64))
+	if err := rn.InitiateHandshake(nonce); err != nil {
+		glog.Errorf("RemoteNode.InitiateHandshake: Error initiating handshake: %v", err)
+		hc.rnManager.Disconnect(rn)
+	}
+	hc.usedNonces.Add(nonce)
+}
+
+// _handleHandshakeCompleteMessage handles HandshakeComplete control messages, sent by RemoteNodes.
+func (hc *HandshakeController) _handleHandshakeCompleteMessage(origin *Peer, desoMsg DeSoMessage) {
+	if desoMsg.GetMsgType() != MsgTypePeerHandshakeComplete {
+		return
+	}
+
+	// Get the handshake information of this peer.
+	remoteNode := hc.rnManager.GetRemoteNodeFromPeer(origin)
+	if remoteNode == nil {
+		return
+	}
+
+	if remoteNode.GetNegotiatedProtocolVersion() == ProtocolVersion2 {
+		if err := hc.handleHandshakeCompletePoSMessage(remoteNode); err != nil {
+			glog.Errorf("HandshakeController._handleHandshakeCompleteMessage: Error handling PoS handshake peer message: %v", err)
+			hc.rnManager.Disconnect(remoteNode)
+			return
+		}
+	}
+	hc.rnManager.ProcessCompletedHandshake(remoteNode)
```

Comment:
nit: nice way to reduce nesting
```suggestion
	if remoteNode.GetNegotiatedProtocolVersion() != ProtocolVersion2 {
		hc.rnManager.ProcessCompletedHandshake(remoteNode)
		return
	}
	
	if err := hc.handleHandshakeCompletePoSMessage(remoteNode); err != nil {
		glog.Errorf("HandshakeController._handleHandshakeCompleteMessage: Error handling PoS handshake peer message: %v", err)
		hc.rnManager.Disconnect(remoteNode)
		return
	}

```